### ISSUE TITLE
Disposability

### ISSUE THREAD

**Issue by oliverkane** [stevegreatrex/ko.plus#12]
I'm looking for documentation about the proper handling of these extensions to avoid memory management leaks in long lived (eg: SPA) type applications.  

If you can point me in the correct spot or provide an example, I could write a wiki page here for you. Just haven't been able to find anything at a glance.

Native ko.computed'd require proper disposal otherwise they remain active even after their bound DOM element has been removed (by default).

**Comment by IanYates** [comment]
I think you've alluded to this with your "by default", but just to clarify...

As far as I know, it's not strictly the case that every ko.computed needs to be disposed.  The browser garbage collector, similar to Java/C#, work on anchoring root references.  This would be objects referenced by "window", which eventually includes DOM elements referencing observables & computeds, etc.  

However if you have some DOM nodes torn down, and those nodes were referencing some view model (and were the only things referencing the VM), and that VM has some observables, a computed, etc, then they'll all get garbage collected for you.

According to the docs...

> dispose() — Manually disposes the computed observable, clearing all subscriptions to dependencies. This function is useful if you want to stop a computed observable from being updated or want to clean up memory for a computed observable that has dependencies on observables that won’t be cleaned.


So I can see merit in having a way to dispose the ko.command (which is the main part of this library that I use), so that it in turn disposes of its canExecute computed, but in practice it hasn't really been a big deal so far, since if the command is no longer applicable, it's always because the view model itself is no longer applicable, because the view that was keeping it around is itself disappearing.

**Comment by oliverkane** [comment]
While not strictly necessary, I a large segment of my population uses an outdated *cough! IE8 cough!* browser, which behaves quite poorly when dealing with nested contexts and compute'ds.  If there's not a way, I may come back to fork the library for you, but I'll sadly not be able to use it in my project without that support.

In particular, I spent some time with a profile before finding this guy:
https://support.microsoft.com/en-us/kb/830555/

Basically, while the GC in most engines understand that a cluster of object have gone out of scope, a single circular reference (in this case computed<-->DOM element) will keep that collection (including the whole VM sometimes) in memory.

**Comment by oliverkane** [comment]
Also, either way, I must say that the ko.command concept is pretty spot on. I stumbled upon this project from http://blog.greatrexpectations.com/ while looking as Swagger.io, and read a few articles.  This paradigm is a much cleaner version of what I'm already doing in a DurandalJS project.  

Thanks for the great work!  I hope I get an opportunity to use this.

**Comment by IanYates** [comment]
Yuk, that's a hard bug to work around in IE 8 :(  I've been lucky enough that I can ask for IE 11, Chrome or FireFox for our work....  The command pattern really is great...  Maybe if we changed it to a pureComputed in this library some of the problems might go away for you? (at the expense of possibly evaluating more often than necessary)

**Comment by stevegreatrex** [comment]
I think it would be a good idea to prefer pureComputed and fallback to
computed so we don't lose support for earlier knockout versions.

**Comment by stevegreatrex** [comment]
Having looked at the guidelines for `pureComputed` I now think that perhaps it isn't appropriate.  If a dependency of `canExecute` changes then it should immediately notify observers.

**Comment by oliverkane** [comment]
Is it possible to rather than implement the disposable method, is it possible for my app to do this manually?  I already have factory methods for ko.computeds so they get disposed properly, I wouldn't mind having such a function, however verbose, around to do the same. I'm just not sure how to get those nested parts into scope to do so without creating more of a mess.

**Comment by stevegreatrex** [comment]
You always have access to the created computed (`ko.command(...).canExecute`) so you could certainly hook up manual disposal through that.

What do your current factory methods for `ko.computed`s look like?

**Comment by oliverkane** [comment]
Actually, that'll work.  I have a method for each of the observable types and I call them similar to how ko.computed is called (same arguments) but I intercept them and save them to an array. Upon the module's disposal, I iterate over these elements, checking for, and calling dispose() on each.

Since I can just access canExecute directly, it would appear that I can do something very similar with that approach.  

Thank you much for your time explaining that to me!
